vm/hash_test.go

```diff
@@ -3,42 +3,650 @@ package vm
+		{`{ a: 1, b: 2 } == [1, "String", true, 2..5]`, false},
+		{`{ a: 1, b: 2 } == Integer`, false},
+		{`{ a: 1, b: 2 } != 123`, true},
+		{`{ a: 1, b: 2 } != "123"`, true},
+		{`{ a: 1, b: 2 } != "124"`, true},
+		{`{ a: 1, b: 2 } != (1..3)`, true},
+		{`{ a: 1, b: 2 } != { a: 1, b: 2 }`, false},
+		{`{ b: 2, a: 1 } != { a: 1, b: 2 }`, false}, // Hash has no order issue
+		{`{ a: 1, b: 2 } != { a: 2, b: 1 }`, true},
+		{`{ a: 1, b: 2 } != { b: 1, a: 2 }`, true},
+		{`{ a: 1, b: 2 } != { a: 1, b: 2, c: 3 }`, true},
+		{`{ a: 1, b: 2 } != { a: 2, b: 2, a: 1 }`, false}, // Hash front key will be overwritten if duplicated
+		{`{ a: [1, 2, 3], b: 2 } != { a: [1, 2, 3], b: 2 }`, false},
+		{`{ a: [1, 2, 3], b: 2 } != { a: [3, 2, 1], b: 2 }`, true}, // Hash of array has order issue
+		{`{ a: 1, b: 2 } != [1, "String", true, 2..5]`, true},
+		{`{ a: 1, b: 2 } != Integer`, true},
+	}
+
+	for i, tt := range tests {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		checkExpected(t, i, evaluated, tt.expected)
+		vm.checkCFP(t, i, 0)
+	}
+}
+
+func TestHashClearMethod(t *testing.T) {
+	input := `
+	{ foo: 123, bar: "test", baz: true }.clear
+	`
+
+	vm := initTestVM()
+	evaluated := vm.testEval(t, input)
+
+	h, ok := evaluated.(*HashObject)
+	if !ok {
+		t.Fatalf("Expect evaluated value to be a hash. got: %T", evaluated)
+	} else if h.length() != 0 {
+		t.Fatalf("Expect length of pairs of hash to be 0. got: %v", h.length())
+	}
+
+	vm.checkCFP(t, 0, 0)
+}
+
+func TestHashClearMethodFail(t *testing.T) {
+	testsFail := []struct {
+		input   string
+		errType string
+		errMsg  string
+	}{
+		{`{ a: 1, b: 2 }.clear(123)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 1"},
+		{`{ a: 1, b: 2 }.clear(true, { hello: "World" })`, ArgumentError, "ArgumentError: Expect 0 argument. got: 2"},
+	}
+
+	for i, tt := range testsFail {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		checkError(t, i, evaluated, tt.errType, tt.errMsg)
+		vm.checkCFP(t, i, 1)
+	}
+}
+
+func TestHashEachKeyMethod(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected []interface{}
+	}{
+		{`
+			{ a: "Hello", b: "World", c: "Goby" }.each_key do |key|
+			  # Empty Block
+			end
+		`, []interface{}{"a", "b", "c"}},
+		{`
+			{ b: "Hello", c: "World", a: "Goby" }.each_key do |key|
+			  # Empty Block
+			end
+		`, []interface{}{"a", "b", "c"}},
+		{`
+			{ b: "Hello", c: "World", b: "Goby" }.each_key do |key|
+			  # Empty Block
+			end
+		`, []interface{}{"b", "c"}},
+		{`
+			arr = []
+			{ a: "Hello", b: "World", c: "Goby" }.each_key do |key|
+			  arr.push(key)
+			end
+			arr
+		`, []interface{}{"a", "b", "c"}},
+	}
+
+	for i, tt := range tests {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		testArrayObject(t, i, evaluated, tt.expected)
+		vm.checkCFP(t, i, 0)
+	}
+}
+
+func TestHashEachKeyMethodFail(t *testing.T) {
+	vm := initTestVM()
+
+	testArgumentError := `
```

Comment:
I think we better not writing test in this way. We should still uses table structure for this like
```go
tests := []struct{
    input string
    expectedErr string
    expectedCfp int
} { ..... }
```